### PR TITLE
[BEAM-1353] Easy PTransform style guide compliance fixes

### PR DESCRIPTION
**Pull request by jkff**
This leaves out the more complex / potentially-controversial ones.

R: @dhalperi

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java [hunk 1]
```diff
@@ -153,11 +153,12 @@ public T extractOutput(TimestampedValue<T> accumulator) {
...
    * @param <V> The value type of the elements being combined.
    */
-  public static <K, V> PTransform<PCollection<KV<K, V>>, PCollection<KV<K, V>>> perKey() {
+  public static <K, V> PerKey<K, V> perKey() {
     return new PerKey<>();
   }
 
-  private static class Globally<T> extends PTransform<PCollection<T>, PCollection<T>> {
+  /** Implementation of {@link #globally()}. */
+  public static class Globally<T> extends PTransform<PCollection<T>, PCollection<T>> {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: I may have missed the relevant section in the PTransform style guide -- why is it good to expose the classes if the user need not further configure them?

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: In other words, why not "principle of least visibility"?

Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: I started a thread about this on beam-dev https://lists.apache.org/thread.html/f448eb9f531bec2c0cd70e790dc8d2198d8a1bf9d2b85f94d9ac904a@%3Cdev.beam.apache.org%3E . There indeed is no section about that, but perhaps my arguments in the thread shall be convincing.

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: I'm not seeing how this discussion is related to that discussion.

It's more internally-evolvable to return PTransform than to return any concrete class.

And the argument about runner intercepting isn't relevant post-FnApi. No runner can inspect the Java classes specifically, only their standard serialized form.

**Comment by jkff** [comment]
I've revised my changes to the "glorified combine" transforms according to the beam-dev thread.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java [hunk 2]
```diff
@@ -297,7 +297,7 @@ public void populateDisplayData(DisplayData.Builder builder) {
   @Test
   public void testWriteUnbounded() {
     PCollection<String> unbounded = p.apply(CountingInput.unbounded())
-        .apply(ToString.of());
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: We decided this one the other way. @eljefe6a

Comment (eljefe6a) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: Are the other violations going to be separate PRs or part of this PR. Should we discuss the method name here or on the list?

Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: I can just take out this commit from the PR if the rest looks fine. The list or the JIRA thread is probably the best place.

Comment (eljefe6a) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java]: Moving this discussion to the list.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java [hunk 3]
```diff
@@ -83,7 +116,7 @@ public LatestFn() {}
 
     @Override
     public TimestampedValue<T> addInput(TimestampedValue<T> accumulator,
-        TimestampedValue<T> input) {
+                                        TimestampedValue<T> input) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: optional style: usually the old indent is preferred

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java [hunk 4]
```diff
@@ -95,16 +128,20 @@ public LatestFn() {}
     }
 
     @Override
-    public Coder<TimestampedValue<T>> getAccumulatorCoder(CoderRegistry registry,
-        Coder<TimestampedValue<T>> inputCoder) throws CannotProvideCoderException {
+    public Coder<TimestampedValue<T>> getAccumulatorCoder(
+        CoderRegistry registry, Coder<TimestampedValue<T>> inputCoder)
+        throws CannotProvideCoderException {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: optional style: usually the throws clause is further indented (I believe)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java [hunk 5]
```diff
@@ -95,16 +128,20 @@ public LatestFn() {}
...
     }
 
     @Override
-    public Coder<T> getDefaultOutputCoder(CoderRegistry registry,
-        Coder<TimestampedValue<T>> inputCoder) throws CannotProvideCoderException {
-      checkState(inputCoder instanceof TimestampedValue.TimestampedValueCoder,
-          "inputCoder must be a TimestampedValueCoder, but was %s", inputCoder);
+    public Coder<T> getDefaultOutputCoder(
+        CoderRegistry registry, Coder<TimestampedValue<T>> inputCoder)
+        throws CannotProvideCoderException {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: (similar throws clause)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sample.java [hunk 6]
```diff
@@ -64,80 +73,74 @@
...
   /////////////////////////////////////////////////////////////////////////////
 
-  /**
-   * A {@link PTransform} that takes a {@code PCollection<T>} and a limit, and
-   * produces a new {@code PCollection<T>} containing up to limit
-   * elements of the input {@code PCollection}.
-   */
-  public static class SampleAny<T> extends PTransform<PCollection<T>, PCollection<T>> {
+  /** Implementation of {@link #any(long)}; do not use directly. */
+  @VisibleForTesting
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sample.java]: It looks like this test is Sample-dependent only for convenience. Is it worth changing the test to avoid the need to make this public?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java [hunk 7]
```diff
@@ -38,74 +37,67 @@ private ToString() {
...
   /**
-   * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
-   * element of the input {@link PCollection} to a {@link String} using the
-   * {@link Object#toString} method.
+   * Transforms each element of the input {@link PCollection} to a {@link String} using the {@link
+   * Object#toString} method.
    */
-  public static PTransform<PCollection<?>, PCollection<String>> elements() {
-    return new SimpleToString();
+  public static Elements elements() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: I'm clearly dumb, because I've asked this question in different flavors a few times, but once more for good measure. Sorry!

The last few commits in this PR left inner transforms generic; why is Elements public? Is it for further customization, or another reason for a difference in visibility?

Comment (eljefe6a) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: No reason. I thought all of the public to private changes were being done in a different PR and all at once.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java (hunk 3 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: I usually apply google-java-format on the changed bits but in this case I guess I forgot.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java (hunk 4 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: I looked at some other instances of a `throws` on a new line in the codebase, and this style seems more common. It is also what `google-java-format` produces.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java (hunk 5 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Latest.java]: Ditto.

**Comment by dhalperi** [comment]
Thanks Eugene. I will merge when green.

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
OK, this failure is pretty clearly a flake, as well as the previous one.

**Comment by dhalperi** [comment]
+1 will merge when i get back. Any other committer welcome to preempt me
